database/src/ops/block.rs

```diff
@@ -1,89 +1,454 @@
+        long_term_weight: block_info.long_term_weight as usize,
+    })
 }
 
-/// TODO
-pub fn get_block_header() {
-    todo!()
+/// Return the top/latest [`ExtendedBlockHeader`] from the database.
+#[doc = doc_error!()]
+#[inline]
+pub fn get_block_extended_header_top(
+    tables: &impl Tables,
+) -> Result<(ExtendedBlockHeader, BlockHeight), RuntimeError> {
+    let height = chain_height(tables.block_heights())?.saturating_sub(1);
+    let header = get_block_extended_header_from_height(&height, tables)?;
+    Ok((header, height))
 }
 
-/// TODO
-pub fn get_block_header_from_height() {
-    todo!()
+//---------------------------------------------------------------------------------------------------- Misc
+/// Retrieve a [`BlockHeight`] via its [`BlockHash`].
+#[doc = doc_error!()]
+#[inline]
+pub fn get_block_height(
+    block_hash: &BlockHash,
+    table_block_heights: &impl DatabaseRo<BlockHeights>,
+) -> Result<BlockHeight, RuntimeError> {
+    table_block_heights.get(block_hash)
 }
 
-/// TODO
-pub fn get_top_block() {
-    todo!()
+/// Check if a block exists in the database.
+#[doc = doc_error!()]
+#[inline]
+pub fn block_exists(
+    block_hash: &BlockHash,
+    table_block_heights: &impl DatabaseRo<BlockHeights>,
+) -> Result<bool, RuntimeError> {
+    table_block_heights.contains(block_hash)
 }
 
-/// TODO
-pub fn get_top_block_hash() {
-    todo!()
+//---------------------------------------------------------------------------------------------------- Tests
+#[cfg(test)]
+#[allow(
+    clippy::significant_drop_tightening,
+    clippy::cognitive_complexity,
+    clippy::too_many_lines
+)]
+mod test {
+    use hex_literal::hex;
+    use pretty_assertions::assert_eq;
+
+    use cuprate_test_utils::data::{block_v16_tx0, block_v1_tx2, block_v9_tx3, tx_v2_rct3};
+
+    use super::*;
+    use crate::{
+        ops::tx::{get_tx, tx_exists},
+        tests::{assert_all_tables_are_empty, tmp_concrete_env},
+        Env,
+    };
+
+    /// Tests all above block functions.
+    ///
+    /// Note that this doesn't test the correctness of values added, as the
+    /// functions have a pre-condition that the caller handles this.
+    ///
+    /// It simply tests if the proper tables are mutated, and if the data
+    /// stored and retrieved is the same.
+    #[test]
+    fn all_block_functions() {
+        let (env, tmp) = tmp_concrete_env();
+        let env_inner = env.env_inner();
+        assert_all_tables_are_empty(&env);
+
+        let mut blocks = [
+            block_v1_tx2().clone(),
+            block_v9_tx3().clone(),
+            block_v16_tx0().clone(),
+        ];
+        // HACK: `add_block()` asserts blocks with non-sequential heights
+        // cannot be added, to get around this, manually edit the block height.
+        for (height, block) in blocks.iter_mut().enumerate() {
+            block.height = height as u64;
+            assert_eq!(block.block.serialize(), block.block_blob);
+        }
```

Comment:
`ops/` tests use real data from `test-utils/` now.

But, we modify the height so `add_block()` doesn't panic - kinda hacky. Could remove the asserts if you'd prefer.

Reply:
I mean the data added isn't 100% accurate anyway - the indexs for outputs will be incorrect so should be fine